### DISCUSSION TITLE
Workaround for ":where" selector for old browsers

### DISCUSSION THREAD

**Discussion by northern-mindset** [csstools/postcss-plugins#509] [Ideas]
Currently global coverage is around 90%: https://caniuse.com/?search=where

But there is no plugin to workaround this selector for old browsers. Is it technically possible? Found this not-ideal suggestion: https://stackoverflow.com/a/71049791/3082178.

**Comment by Antonio-Laguna** [comment]
@anantakrishna thanks for asking!

The issue is what they outline in the comments. The main feature of `:where` is to achieve `0` specificity which can't be emulated with plain CSS as far as I know. We can increase specificity (as we've done for CSS Layers) but can't lower it.

**Reply by northern-mindset** [reply]
Thanks for a quick reply!

Yes, I understand the issue. Don't you consider that emulating `:where` with increased specificity still could be beneficial? At least there may be a chance of proper styling in the old browsers instead of completely unstyled HTML?

Examples:

- https://github.com/tailwindlabs/tailwindcss/discussions/8001
- https://github.com/facebook/docusaurus/issues/7617

**Reply by Antonio-Laguna** [reply]
@anantakrishna that's a good point!

I don't think simply desugaring is the way to go here as the fallback has the side effect of just increasing specificity which means it loses the point. Such a plugin _could_ be easily created but I don't think we would bundle it as is into preset-env given it would not be faithful to the spec since specificity wouldn't be preserved.

Now, discussing with @romainmenke, we should be able to reuse part of what was done for Cascade Layers as `:where` is a special cascade layer. That would, however, lead to the same sort of output we get out of Cascade Layers which means something like this.

Taking your example:

```pcss
:where(header, main, footer) p:hover {
  color: red;
  cursor: pointer;
}

p {
  /* Unrelated */
  color: blue;
}
```

Would decompress to:

```pcss
header p:hover,
main p:hover,
footer p:hover {
  color: red;
  cursor: pointer;
}

p:not(#\#) {
  color: blue;
}
```

This quickly explodes as your CSS grows and more and more `:where` is used which is something that makes us a bit nervous. It can be done but not something we're aiming for in the immediate future given the implications it can have.

--- 

All of that said, I think it should be safer to desugar yourself instead of leveraging `:where` if you're giving support for older browsers that do not support it.

**Reply by northern-mindset** [reply]
Thank you for giving it a thought.

The point is that the code is not of mine, it's from TailwindCSS Typography plugin (https://github.com/tailwindlabs/tailwindcss/discussions/8001). Anyhow, as of now I am going to resort to just ignoring the issue in the old browsers.

**Reply by andersk** [reply]
Similarly, Astro automatically inserts `:where()` to enable scoped CSS in components without interfering with specificity.

- https://docs.astro.build/en/guides/styling/#scoped-styles
- https://github.com/withastro/rfcs/pull/63
- https://github.com/withastro/astro/pull/4024
